### PR TITLE
[DPE-4179] Add integration test for upgrades and rollbacks

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
We are missing integration tests for upgrades and rollbacks

## Solution
Add tests

### CODE REVIEW

File: tests/integration/test_log_rotation.py [hunk 1]
```diff
@@ -24,9 +26,9 @@
 
 METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
 
-MYSQL_APP_NAME = "mysql-k8s"
+MYSQL_APP_NAME = MYSQL_DEFAULT_APP_NAME
 MYSQL_ROUTER_APP_NAME = "mysql-router-k8s"
```
Comment (taurus-forever) [inline tests/integration/test_log_rotation.py:30]: why MYSQL_ROUTER_APP_NAME was not moved to helpers? It looks half-baked.

**Review by taurus-forever — APPROVED** [review]
LGTM. One nit.

File: tests/integration/helpers.py [hunk 2]
```diff
@@ -431,3 +438,205 @@ async def get_tls_certificate_issuer(
...
+                    )
+                    numbers = {n for n in range(1, max_written_value)}
+                    assert (
+                        numbers <= all_written_values
+                    ), f"Missing numbers in database for unit {unit.name}"
+
+                    last_max_written_value = max_written_value
+
+
+async def get_workload_version(ops_test: OpsTest, unit_name: str) -> None:
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/helpers.py]: addressed in 6e00cbc389baf29f68c607c218033729e8c37c21

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -431,3 +438,205 @@ async def get_tls_certificate_issuer(
...
+    return leader_unit
+
+
+def get_juju_status(model_name: str) -> str:
+    """Return the juju status output.
+
+    Args:
+        model_name: The model for which to retrieve juju status for
+    """
+    return subprocess.check_output(["juju", "status", "--model", model_name]).decode("utf-8")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/helpers.py:643]: good call, addressed in 6e00cbc389baf29f68c607c218033729e8c37c21

File: tests/integration/juju_.py [hunk 4]
```diff
@@ -3,12 +3,26 @@
...
 _libjuju_version = importlib.metadata.version("juju")
+juju_major_version = int(_libjuju_version.split(".")[0])
 is_3_1_or_higher = (
     len(_libjuju_version.split(".")) >= 2
     and int(_libjuju_version.split(".")[0]) >= 3
     and int(_libjuju_version.split(".")[1]) >= 1
 )
 
 is_3_or_higher = int(_libjuju_version.split(".")[0]) >= 3
+
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/juju_.py]: :+1:, changed in 6e00cbc389baf29f68c607c218033729e8c37c21

File: tests/integration/test_log_rotation.py (hunk 1 above)
Comment (shayancanonical) [reply tests/integration/test_log_rotation.py]: it does indeed look half baked, consistency is important. moved to helpers in 6e00cbc389baf29f68c607c218033729e8c37c21

**Pull request merged**